# 課題5 テキストデータ分析
配点
- Q1, 1P
- Q2.1, 0.5P
- Q2.2, 0.5P
- Q2.3, 1P
- Q3.1, 1P
- Q3.2, 2P
- Q3.3, 2P
- Q4, 2P
## Q1

$n$-次元ベクトル空間における、任意の2つのベクトル、$ \vec{x}=(x_1,x_2,...,x_n)$、$\vec{y}=(y_1,y_2,...,y_n)$、の間のcos類似度 $cos(X,Y)$ は以下のように定義されます。

$cos(X,Y) =  \frac{\vec{x}\cdot \vec{y}}{\|x\|_2\|y\|_2}=\frac{\Sigma^{n}_{i=1}{x_{i}y_{i}}}{\sqrt{\Sigma^{n}_{i=1}x_{i}^{2}}\sqrt{\Sigma^{n}_{i=1}y_{i}^{2}}}$

入力ベクトル $\vec{x}$, $\vec{y}$ をそれぞれ`NumPy` の配列として引数で受け取り、それらのベクトル間のcos類似度を計算して返す関数 `cos_sim` を完成させてください。

In [2]:
import numpy as np

In [3]:
def cos_sim(vec1, vec2):
    return # (vec1とvec2の内積)/np.sqrt(vec1の要素の二乗和*vec2の要素の二乗和)

`cos_sim()`関数が完成したら、以下のセルを実行して動作を確認してください。`-1.0, 1.0, 0.0`がそれぞれのcos類似度です。

In [4]:
print(cos_sim(np.array([1,0]), np.array([-1,0])))
print(cos_sim(np.array([1,0]), np.array([1,0])))
print(cos_sim(np.array([1,0]), np.array([0,-1])))

-1.0
1.0
0.0


## Q2
"course_list.csv"ファイルには以下のように各行に講義名のテキストデータが含まれています。以下では、このファイルを読み込み、各講義のベクトルを作成し、講義間の類似度を求めるコードを実装します。
```Python
## course_list.csvファイル
...
数理科学基礎
物性化学
政治Ⅰ
身体運動・健康科学実習Ⅰ（ﾒﾃﾞｨｶﾙｹｱ1）
スペイン語テクスト分析Ⅰ
英語二列Ｗ（ALESS）
応用倫理学
...
```
まず必要なモジュールをimportします。

In [5]:
import json
import csv
import numpy as np
import pandas as pd

### Q2.1
"course_list.csv"ファイルを1行ずつ**順番に**読み込み、その各行を要素とするリストを作成して返す`course_list`関数を完成さてください。作成されたリストは変数`courses`で受け取ります。以降の処理では、リスト`courses`のインデックスをその要素（講義名）のIDとして扱います。


`course_list`関数が完成したら、以下のセルを実行して動作を確認してください。リストの長さ（講義の数）は`552`、講義"計算機プログラミング
"のID（リスト`courses`のインデックス）は`198`となります。

In [6]:
def course_list():
    lst=[]
    with open('course_list.csv', 'r',  encoding="utf-8") as f:
        dataReader = csv.reader(f)
        for row in dataReader:
            # lstにrow[0]を追加
    return lst
courses = course_list()
print(len(courses))
print(course_list().index('計算機プログラミング'))

552
198


### Q2.2
"keyword_list.csv"ファイルには以下のように各行に1単語が含まれています。
```Python
## keyword_list.csvファイル
...
外国語
日本語
力学
身体運動
微分積分学
ロシア語
熱力学
線型代数学
イタリア語
物性
...
```
"keywor_list.csv"ファイルを1行ずつ**順番に**読み込み、その各行を要素とするリストを作成して返す`vocab_list`関数を完成さてください。作成されたリストは変数`vocab`で受け取ります。以降の処理では、リスト`vocab`のインデックスをその要素（単語）のIDとして扱います。リスト`vocab`は以降の処理における語彙となります。

`vocab_list`関数が完成したら、以下のセルを実行して動作を確認してください。リストの長さ（単語の数）は`243`、単語"プログラミング"のID（リスト`vocab`のインデックス）は`145`となります。

In [7]:
def vocab_list():
    lst = []
    with open('keyword_list.csv', 'r',  encoding="utf-8") as f:
        dataReader = csv.reader(f)
        for row in dataReader:
            # lstにrow[0]を追加
    return lst
vocab = vocab_list()
print(len(vocab))
print(vocab.index('プログラミング'))
print(vocab.index('英語'))

243
145
0


### Q2.3
リスト`courses`と`vocab`を引数で受け取り、単語のID（リスト`vocab`のその単語のインデックス）をキー、その単語のDF（Document Frequency)を値とする辞書を作成して返す`count_df`関数を作成してください。作成された辞書は変数`df`で受け取ります。この場合、ある単語のDFはその単語を講義名に含む講義数に対応します。

`count_df`関数が完成したら、以下のセルを実行して動作を確認してください。辞書の長さは`243`、単語"プログラミング"（IDは`145`）のDFは`2`、単語"英語"（IDは`0`）のDFは`12`となります。

In [8]:
# df
def count_df(courses, vocab):
    dic={}
    for i in range(len(vocab)):
        for course in courses:
            if # courseがvocab[i]を含むという条件:
                # dic[i]を1つ増やす 
    return dic
df = count_df(courses, vocab)
print(len(df))
print(vocab.index('プログラミング'), df[vocab.index('プログラミング')])
print(vocab.index('英語'), df[vocab.index('英語')])

243
145 2
0 12


## Q3
### Q3.1
リスト`vocab`の各単語を次元とする講義ベクトルを考えます。講義ベクトルの長さはリスト`vocab`の長さと等しく、リスト`vocab`のインデックス`i`の単語`vocab[i]`が講義名に含まれる時、講義ベクトルの`i`番目の要素は`1`、含まれなければ`0`とします。

以下では、リスト`courses`と`vocab`を引数で受け取り、リスト`courses`の各講義のベクトルを行、リスト`vocab`の各単語を列とした`NumPy`の行列を作成して返す`lec_word_matrix`関数を完成させてください。作成した講義-単語行列は、講義（行）の講義名に単語（列）が含まれていれば、その要素が1であるような行列です。

`lec_word_matrix`関数が完成したら、以下のセルを実行して動作を確認してください。作成した講義-単語行列の要素の値の和は`2287`となります。

In [9]:
# document-term matrix
def lec_word_matrix(courses, vocab):
    mat = np.zeros((# 講義数, 授業数))
    for i in range(len(courses)):
        for j in range(len(vocab)):
            if # courses[i]がvocab[i]を含むという条件（ヒント in ）:
                # mat[i, j]に1を代入
    return mat
np.sum(lec_word_matrix(courses, vocab))

2287.0

### Q3.2
Q3.1で作成した講義-単語行列の各要素を、その講義の講義名に単語が含まれるか否かの1or0ではなく、その講義の講義名に単語が何回含まれるか（TF: Term Frequency)で表した行列を作成して返す`lec_word_tf_matrix`関数を完成させてください。

In [10]:
# document-term matrix
def lec_word_tf_matrix(courses, vocab):
    mat = np.zeros((# 講義数, 授業数))
    for i in range(len(courses)):
        for j in range(len(vocab)):
            if # courses[i]がvocab[i]を含むという条件:
                # mat[i, j]にcourses[i]に含まれるvocab[j]の数を代入（ヒント count())
    return mat
np.sum(lec_word_tf_matrix(courses, vocab))

2366.0

`lec_word_tf_matrix`関数が完成したら、以下のセルを実行して動作を確認してください。作成した講義-単語行列の要素の値の和は`2366`となります。

講義「自然科学ゼミナール（情報科学）」には「科学」が2回含まれるので対応する行列の要素の値は2となります。

In [11]:
lec_word_tf_matrix(courses, vocab)[courses.index('自然科学ゼミナール（情報科学）'), vocab.index('科学')]

2.0